### PR TITLE
protocol-dts-generator for devtools-protocol typescript types

### LINKED ISSUE #55: FAQ
Just wrote a few FAQ entries that I'll add to the site soon.   Capturing here for the moment:

#### How is the protocol defined?  
The canonical protocol definitions live in the Chromium source tree: ([browser\_protocol.json](https://chromium.googlesource.com/chromium/src/+/master/third_party/WebKit/Source/core/inspector/browser_protocol.json) and [js\_protocol.json](https://chromium.googlesource.com/v8/v8/+/master/src/inspector/js_protocol.json)). They are maintained manually by the DevTools engineering team. These files are mirrored (hourly) on GitHub [in the devtools-protocol repo](https://github.com/ChromeDevTools/devtools-protocol/tree/master/json). 

The declarative protocol definitions are used across tools. Within Chromium, a binding layer is created for the Chrome DevTools to interact with, and separately the protocol is used for [Chrome Headless's C++ interface](https://chromium.googlesource.com/chromium/src/+/lkgr/headless/README.md#client_devtools-api).

#### What's the protocol\_externs file?  
It's created via [generate\_protocol\_externs.py](https://cs.chromium.org/chromium/src/third_party/WebKit/Source/devtools/scripts/build/generate_protocol_externs.py) and useful for tools using closure compiler. The TypeScript story is [here](https://github.com/ChromeDevTools/devtools-protocol/issues/18).

#### Are the HTTP endpoints documented?  
Not yet. See bugger-daemon's [third-party docs](https://github.com/buggerjs/bugger-daemon/blob/master/README.md#api). See also the [endpoints implementation](https://cs.chromium.org/search/?q=f:devtools_http_handler.cc+%22command+%3D%3D+%22&sq=package:chromium&type=cs) in Chromium.  `/json/protocol` was added in Chrome 60.

#### How do I access the browser target?  
The endpoint is exposed as `webSocketDebuggerUrl` in `/json/version`. Note the `browser` in the URL, rather than `page`. If Chrome was launched with `--remote-debugging-port=0` and chose an open port, the browser endpoint is written to both stderr and the `DevToolsActivePort` file in browser profile folder.

### PR DESCRIPTION
**Pull request by nojvek**
Fixes #55 (typescript definitions)

## TODO List
 - [x] Create a protocol-dts-generator in scripts.
 - [x] Generator invoked on publish via "prepublishOnly" command
 - [x] `types` entry in package.json 
 - [x] Generated typescript definition: https://github.com/nojvek/devtools-protocol/blob/9edcecb0dcca0c2e654db94a3ba1fbf3eb94fd7e/types/protocol.d.ts
 - [ ] Cleanup generator code
 - [ ] Ensure protocol.json files are typechecked with protocol-schema.d.ts
 - [ ] Some documentation on how to use .d.ts

### CODE REVIEW

**Comment by nojvek** [comment]
@paulirish ^

**Comment by JoelEinbinder** [comment]
This is very cool. I tried to use it to add types to puppeteer and ran into trouble with events. Having a separate method per event makes it difficult to work with. Id expect the domains to extend Node's EventEmitter and have events defined like
```ts
on(event: 'TargetInfoChanged', listener: (params: Target.TargetInfoChangedEvent) => void): this;
```

Also every method is marked as optional on every domain. Maybe this should correspond to whether or not the method is experimental? Or have some other indication of experimental vs stable methods.

**Comment by nojvek** [comment]
@JoelEinbinder done.

Yeah I left it as onXYZ because typescript would give better code completion, but they've fixed that since then.

Happy to take feedback on what you'd like the ideal interface to be.

do you simply want the interface to be .

```
send(command: 'Domain.xyz', params: SomeRequest): Promise<SomeResponse>
on(event: 'Domain.xyz', listener: (params: SomeEvent) => void): void
```

File: scripts/protocol-dts-generator.ts [hunk 1]
```diff
@@ -0,0 +1,228 @@
+import * as fs from 'fs'
+import * as path from 'path'
+import {IProtocol, Protocol as P} from './protocol'
```
Comment (paulirish) [inline scripts/protocol-dts-generator.ts]: `protocol-schema`, now, eh?

**Comment by brendankenny** [comment]
FWIW I've found the overloaded `on()` doesn't play well with using these types from js (using jsdocs), but OTOH it is the best way to get type inferencing from the compiler (to avoid given manual types to event responses). I'm playing with a different approach for Lighthouse since we have places that deal with consuming and producing (or at least re-playing) protocol methods. If I get it working I can make a proposal to combine it with this work so both options are supported.

All the info we need is in here, though, and as long as all the method definitions are on `DomainApi` or whatever (like you have now), we're good.

**Comment by nojvek** [comment]
@JoelEinbinder  @brendankenny  Please let me know your preferences for the final interface and I'll make the changes.

**Comment by paulirish** [comment]
@JoelEinbinder you happy with the look of this now?

**Comment by paulirish** [comment]
@nojvek the remaining TODOs are some good stuff. do you want to do those in this PR or the next one? Ideally we wouldn't delay it too much.

**Comment by nojvek** [comment]
I’m happy it being a v1 and work on documentaion and protocol steps in next PR

How does the publish of npm modules work? I imagine some bot runs the script?

**Comment by paulirish** [comment]
@nojvek Yup. Done by a bot every hour.

**Comment by themaxdavitt** [comment]
I created a PR to ensure the `*_protocol.json` files are typechecked with `protocol-schema.d.ts` (from the TODO list): https://github.com/ChromeDevTools/devtools-protocol/pull/211

**Pull request merged**